File: leetcode/py/p423-reconstruct-original-digits-from-english.py
# LeetCode 423. Reconstruct original digits from English.
import collections


class Solution:
    def originalDigits(self, s: str) -> str:
        num_0 = s.count("z")
        num_2 = s.count("w")
        num_4 = s.count("u")
        num_5 = s.count("f") - num_4
        num_6 = s.count("x")
        num_8 = s.count("g")
        l = []
        l.append("0" * num_0)
        l.append("1" * (s.count("o") - num_0 - num_2 - num_4))
        l.append("2" * num_2)
        l.append("3" * (s.count("h") - num_8))
        l.append("4" * num_4)
        l.append("5" * num_5)
        l.append("6" * num_6)
        l.append("7" * (s.count("s") - num_6))
        l.append("8" * num_8)
        l.append("9" * (s.count("i") - num_5 - num_6 - num_8))
        return "".join(l)


class Solution2:
    def originalDigits(self, s: str) -> str:
        counts = collections.Counter()
        for c in s:
            counts[c] += 1
        search_things = (
            ("z", "ro", "0"),
            ("w", "o", "2"),
            ("u", "for", "4"),
            ("o", "", "1"),
            ("r", "", "3"),
            ("f", "i", "5"),
            ("x", "si", "6"),
            ("s", "", "7"),
            ("g", "i", "8"),
            ("i", "", "9"),
        )
        l = []
        for st in search_things:
            count = counts[st[0]]
            if count > 0:
                l.extend([st[2]] * count)
                for c in st[1]:
                    counts[c] -= count
        l.sort()
        return "".join(l)


class Solution3:
    def originalDigits(self, s: str) -> str:
        counts = collections.Counter()
        for c in s:
            counts[c] += 1

        search_things = (
            ("z", "zero", "0"),
            ("w", "two", "2"),
            ("u", "four", "4"),
            ("o", "one", "1"),
            ("r", "three", "3"),
            ("f", "five", "5"),
            ("x", "six", "6"),
            ("s", "seven", "7"),
            ("g", "eight", "8"),
            ("i", "nine", "9"),
        )
        l = []
        for st in search_things:
            count = counts[st[0]]
            if count > 0:
                l.extend([st[2]] * count)
                for c in st[1]:
                    counts[c] -= count
        l.sort()
        return "".join(l)


if __name__ == "__main__":

    solution = Solution()
    tests = (
        ("owoztneoer", "012"),
        ("fviefuro", "45"),
        ("nineeightsevensixfivefourthreetwoonezero", "0123456789"),
        ("zero", "0"),
        ("zerozero", "00"),
        ("one", "1"),
        ("two", "2"),
        ("three", "3"),
        ("four", "4"),
        ("five", "5"),
        ("six", "6"),
        ("seven", "7"),
        ("eight", "8"),
        ("nine", "9"),
    )

    for test in tests:
        s = test[0]
        exp = test[1]
        ans = solution.originalDigits(s)
        if exp != ans:
            print(f"FAIL. s: {s}, exp: {exp}, ans: {ans}")

File: leetcode/cpp/p423-reconstruct-original-digits-from-english.cpp
// LeetCode 423. Reconstruct original digits from English.
#include <algorithm>
#include <iostream>

class Solution {
public:
    std::string originalDigits(const std::string& s)
    {
        size_t letter_counts[127] = { 0 };
        for (const auto c : s) {
            ++letter_counts[static_cast<int>(c)];
        }
        const auto num_0 = letter_counts['z'];
        const auto num_2 = letter_counts['w'];
        const auto num_4 = letter_counts['u'];
        const auto num_6 = letter_counts['x'];
        const auto num_8 = letter_counts['g'];

        const auto num_1 = letter_counts['o'] - num_0 - num_2 - num_4;
        const auto num_3 = letter_counts['h'] - num_8;
        const auto num_5 = letter_counts['f'] - num_4;
        const auto num_7 = letter_counts['s'] - num_6;

        const size_t ret_size = letter_counts['o']
            + letter_counts['i']
            + letter_counts['h']
            + letter_counts['s']
            - letter_counts['g']
            - letter_counts['x'];

        std::string ret(ret_size, '9');
        auto it = ret.begin();
        it = std::fill_n(it, num_0, '0');
        it = std::fill_n(it, num_1, '1');
        it = std::fill_n(it, num_2, '2');
        it = std::fill_n(it, num_3, '3');
        it = std::fill_n(it, num_4, '4');
        it = std::fill_n(it, num_5, '5');
        it = std::fill_n(it, num_6, '6');
        it = std::fill_n(it, num_7, '7');
        it = std::fill_n(it, num_8, '8');
        return ret;
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        std::string s;
        std::string exp;
    };
    const TestCase testCases[] = {
        { "owoztneoer", "012" },
        { "fviefuro", "45" },
        { "nineeightsevensixfivefourthreetwoonezero", "0123456789" },
        { "zero", "0" },
        { "zerozero", "00" },
        { "one", "1" },
        { "two", "2" },
        { "three", "3" },
        { "four", "4" },
        { "five", "5" },
        { "six", "6" },
        { "seven", "7" },
        { "eight", "8" },
        { "nine", "9" },
    };

    Solution solution;
    for (const auto& tc : testCases) {
        const auto ans = solution.originalDigits(tc.s);
        if (tc.exp != ans) {
            std::cout << "FAIL. " << __FUNCTION__ << "(s: " << tc.s << ")"
                      << ", exp: " << tc.exp
                      << ", ans: " << ans << "\n";
        }
    }
    return 0;
}
